### ISSUE TITLE
Compute lag in custom loss function

### ISSUE THREAD

**Issue by Vidaringa** [rstudio/keras#480]
I wan't to make a custom loss function where I compute the lag of y_pred or y_true. From here: [#257](https://github.com/rstudio/keras/issues/257) I read that I can't convert a tensor to an R object, which I would have to do to use the lag function in R. Taking a look at the backend functions I can't see anyone which compute the lag for me (e.g. k_lag).

Is there any way for me to compute the lag of y_true or y_pred?

**Comment by skeydan** [comment]
Hi, not entirely sure what you want to do, but assuming you want to compare the respective lags (diffs from one point in time to the next) between y_true and y_pred, you could do something like this (simple example with just 2 axes, a batch axis and a feature axis):

```
library(keras)

loss <- function(y_true, y_pred) {
  lag_y_true <- y_true[ , 2:dim(y_true)[2]] - y_true[ , 1:(dim(y_true)[2] - 1)]
  lag_y_pred <- y_pred[ , 2:dim(y_true)[2]] - y_pred[ , 1:(dim(y_true)[2] - 1)]
 # assuming you want to subtract the lags and average them, but could do other things too 
  k_mean(lag_y_pred - lag_y_true)
}

# test
sess <- k_get_session()
ys <- k_constant(c(1,2,3,4,5,6), shape = c(2,3))
yhats <- k_constant(c(1,3,5,7,9,11), shape = c(2,3))
sess$run(ys)
sess$run(yhats)
sess$run(loss(ys, yhats))
```

**Comment by Vidaringa** [comment]
Hi, thank you for your answer. I should have been more explicit about what I'm trying to do. But your solution could help me.

What I'm trying to do is to build a forecasting model. Quite often the forecast is just the true value shifted by one. So I want to adjust the mean absolute error by penalising it if the forecast at t+1 is the same (or really close) to the true value at time t.

I thought of something like this:

customLoss = mean(abs(y_true - y_pred)) + 1/(y_true - lag_y_pred)).

I'm not sure if this will work but I'm experimenting and I need to be able to compute the lag for y_pred.

**Comment by skeydan** [comment]
Ah ok thanks, I see! I think in my code above I was actually doing diff not lag! But as the diff presupposes the lag, yes, I think this should be useful for your case.
